arch/lkl/kernel/cpu.c

```diff
@@ -103,6 +103,7 @@ int lkl_cpu_get(void)
 		__cpu_try_get_unlock(ret, 0);
 		lkl_ops->sem_down(cpu.sem);
 		ret = __cpu_try_get_lock(0);
+		cpu.sleepers--;
```

Comment:
This doesn't look quite right.  `cpu.sleepers` is protected by the mutex, but you are (non-atomically) decrementing it even if you fail to acquire the lock.  If `__cpu_try_get_lock` fails, this is now a data race.

Reply:
Its not because the lock hasn't been released yet.

Reply:
Acquires lock here:

https://github.com/lsds/lkl/blob/upstream-refactor/arch/lkl/kernel/cpu.c#L63

Exits here without unlocking:

https://github.com/lsds/lkl/blob/upstream-refactor/arch/lkl/kernel/cpu.c#L71

It's not til we come back around and do the "try unlock" that it releases the lock:

https://github.com/lsds/lkl/blob/upstream-refactor/arch/lkl/kernel/cpu.c#L82

Quite the code, eh?

Reply:
Right, `__cpu_try_get_lock()` always acquires the lock (unless we're shutting down?)

Reply:
(i.e. the`cpu.lock` that protects the CPU lock state, not the CPU lock itself.)

Reply:
I was tempted to refactor to be more clear, but decided "if it works, best not to touch it".

Reply:
Well, touch it only to fix the bug.

Reply:
When testing this fix, what differences did you see?

Reply:
After a offline discussion with @davidchisnall, he identified a race condition with sleeper decrement that could happen if the __cpu_try_get_lock call returned -2 which happens on shutdown.

I've pushed an update to address that.

Reply:
@prp #384 goes away for "surprising result".

I've been having a hard time triggering bugs so, I can say when testing I didnt get any hangs. Not that that means anything was fixed as I was having a hard time triggering hangs anyway.

Reply:
@SeanTAllen the big question is how this will fare in the CI pipeline, as the pipeline seems to be triggering hangs more consistently. Do you want to create an SGX-LKL PR thats bumps LKL?

Reply:
There is one's to test.

It had several failures of the sort we've been seeing so there is definitely more to find.

I found this while looking for the "hangs calling umount on shutdown" bug that I've found. I still think the mostly likely cause for that hang is with locks related to lkl_syscall and plan on returning to looking at that more until such time as I have proof that it doesn't exist. 

I doubt this would fix that, but I'm not sure.

Reply:
@prp I'm also going to spend more time looking at the shutdown logic in cpu.c as a number of hangs happen during shutdown. It's complicated code and there might be a bug hiding in there.